components/inputs/input-radio-group.js

```diff
@@ -0,0 +1,239 @@
+		}
+		if (changedProperties.has('required')) {
+			this.#recalculateState(true);
+		}
+	}
+
+	focus() {
+		const radios = this.#getRadios();
+		if (radios.length === 0) return;
+		let firstFocusable = null;
+		let firstChecked = null;
+		radios.forEach(el => {
+			if (firstFocusable === null && !el.disabled) firstFocusable = el;
+			if (firstChecked === null && el._checked) firstChecked = el;
+		});
+		const focusElem = firstChecked || firstFocusable;
+		focusElem.focus();
+		setTimeout(() => focusElem.focus()); // timeout required when following link from form validation
+	}
+
+	#labelId = getUniqueId();
+
+	async #doUpdateChecked(newChecked, doFocus, doDispatchEvent) {
+		const radios = this.#getRadios();
+		let prevChecked = null;
+		radios.forEach(el => {
+			if (el._checked) prevChecked = el;
+		});
+		if (prevChecked === newChecked) return;
+
+		newChecked._checked = true;
+		if (prevChecked !== null) {
+			prevChecked._checked = false;
+		}
+		this.#recalculateState(true);
+
+		if (doDispatchEvent) {
+			this.dispatchEvent(new CustomEvent('change', {
+				bubbles: true,
+				composed: true,
+				detail: {
+					value: newChecked.value,
+					oldValue: prevChecked?.value
+				}
+			}));
+		}
+
+		if (doFocus) {
+			await newChecked.updateComplete; // wait for tabindex to be updated
+			newChecked.focus();
+		}
+	}
+
+	#getRadios() {
+		const elems = this.shadowRoot?.querySelector('slot')?.assignedElements();
+		if (!elems) return [];
+		return elems.filter(el => el.tagName === 'D2L-INPUT-RADIO');
+	}
+
+	#handleClick(e) {
+		if (e.target.tagName !== 'D2L-INPUT-RADIO') return;
+		if (e.target.disabled) return;
+		this.#doUpdateChecked(e.target, true, true);
+		e.preventDefault();
+	}
+
+	#handleKeyDown(e) {
+		if (e.target.tagName !== 'D2L-INPUT-RADIO') return;
+
+		const isRtl = (getComputedStyle(this).direction === 'rtl');
+		let newOffset = null;
+		if (e.key === ' ') {
+			newOffset = 0;
+		} else if (e.key === 'ArrowUp' || (!isRtl && e.key === 'ArrowLeft') || (isRtl && e.key === 'ArrowRight')) {
+			newOffset = -1;
+		} else if (e.key === 'ArrowDown' || (!isRtl && e.key === 'ArrowRight') || (isRtl && e.key === 'ArrowLeft')) {
+			newOffset = 1;
+		}
+
+		if (newOffset === null) return;
+
+		const radios = this.#getRadios().filter(el => !el.disabled || el._checked);
+		let checkedIndex = -1;
+		let firstFocusableIndex = -1;
+		radios.forEach((el, i) => {
+			if (el._checked) checkedIndex = i;
+			if (firstFocusableIndex < 0 && !el.disabled) firstFocusableIndex = i;
+		});
+		if (checkedIndex === -1) {
+			if (firstFocusableIndex === -1) return;
+			checkedIndex = firstFocusableIndex;
+		}
+
+		const newIndex = (checkedIndex + newOffset + radios.length) % radios.length;
+		this.#doUpdateChecked(radios[newIndex], true, true);
+
+		e.preventDefault();
+	}
+
+	#handleRadioChecked(e) {
+		if (e.detail.checked) {
+			this.#doUpdateChecked(e.target, false, false);
+		} else {
+			e.target._checked = false;
```

Comment:
This gets called when a consumer programatically unchecks a `<d2l-input-radio>`, which can actually result in nothing being checked. This matches native radio group behaviour, but I'm tempted to make this impossible (i.e. remove this portion of code).

I'm worried that consumers will not make a radio group required, have one checked by default and then not expect it to be possible for the group to not have a value. Which would be true except if someone hacks it in the dev console. Thoughts?

Reply:
What if they have some type of reset on the page - like if the user can reset the form? Or if they've got this inside a dialog and they are trying to re-used the dialog and they want to reset it (tho in the past we've suggested recreating the dialog to hard reset the dialog content)?

Reply:
That's what I'm leaning towards, but I could be convinced the other way.